src/utils/default-plugin.ts

```diff
@@ -1,14 +1,44 @@
 import { InputOptions, Plugin } from '../rollup/types';
 import error from './error';
-import { lstatSync, readdirSync, readFileSync, realpathSync } from './fs'; // eslint-disable-line
+import { lstatSync, readdirSync, readFile, realpathSync, stat, Stats } from './fs'; // eslint-disable-line
 import { basename, dirname, isAbsolute, resolve } from './path';
 
+interface LoadMemoization {
+	[key: string]: {
+		mtimeMs: number;
+		size: number;
+		source: string;
+	};
+}
+
+const loadMemoization: LoadMemoization = Object.create(null);
+
 export function getRollupDefaultPlugin(options: InputOptions): Plugin {
 	return {
 		name: 'Rollup Core',
 		resolveId: createResolveId(options),
 		load(id) {
-			return readFileSync(id, 'utf-8');
+			// In order to avoid too much I/O `load` by default will
+			// only stat files, and do a full read if it doesn't already
+			// hold a files contents in memory.
+			const memoized = loadMemoization[id];
+			if (memoized) {
+				return new Promise<Stats>((resolve, reject) =>
+					stat(id, (err, stats) => (err ? reject(err) : resolve(stats)))
+				).then(({ mtimeMs, size }) => {
+					if (mtimeMs === memoized.mtimeMs && size === memoized.size) return memoized.source;
+					return new Promise<string>((resolve, reject) =>
+						readFile(id, 'utf-8', (err, source) => (err ? reject(err) : resolve(source.toString())))
+					).then(source => {
+						loadMemoization[id] = { mtimeMs, size, source };
+						return source;
+					});
+				});
+			}
+
+			return new Promise<string>((resolve, reject) =>
```

Comment:
Sorry to be a little nit-picky here but as far as I can see, the current implementation will still not memoize anything here unless we add a way for ids which are not yet in `loadMemoization` to be actually added to `loadMemoization`. On a positive side, the memory cost was probably not that high yet :)
Maybe we can restructure the code a little so that the code to read a file and update the memoization is only written once?

I guess that is the pity with things that are hard to test in an E2E test. If we ever add smaller integration tests, this feature would be a prime candidate.